rasterio/_io.pxd

```diff
@@ -42,206 +42,36 @@ ctypedef np.float64_t DTYPE_FLOAT64_t
 cdef bint in_dtype_range(value, dtype)
 
 
-cdef int io_ubyte(
+cdef int io_band(
         GDALRasterBandH band,
         int mode,
         int xoff,
         int yoff,
         int width,
         int height,
-        np.uint8_t[:, :] buffer)
+        object buffer)
```

Comment:
@perrygeo @brendan-ward I'm proposing in this PR to *not* convert the input array to a typed memoryview as we were doing before, but just take the array and then get a `void *` to its bytes using `np.PyArray_DATA()`, which is what Cython does. This allows us to eliminate hundreds of lines of mostly redundant code and hasn't impacted peformance (I timed `rio-convert` for this branch and for master, and there's no significant change -- this one is a little faster).

This *does* change our private API: it only accepts Numpy arrays now, but we were never passing anything else.